.circleci/config.yml

```diff
@@ -0,0 +1,201 @@
+      - run: |
+          npm run lint
+          npm run flow -- check
+          npm test -- --maxWorkers=1
+
+  test-node-6:
+    working_directory: ~/react-native
+    docker:
+      - image: circleci/node:6.11.0
+    steps:
+      - checkout
+      - run: npm install --no-package-lock
+      - run: |
+          npm run lint
+          npm run flow -- check
+          npm test -- --maxWorkers=1
+
+  test-node-4:
+    working_directory: ~/react-native
+    docker:
+      - image: circleci/node:4.8.4
+    steps:
+      - checkout
+      - run: npm install --no-package-lock
+      - run: |
+          npm run lint
+          npm run flow -- check
+          npm test -- --maxWorkers=1
+
+  build-js-bundle:
+    working_directory: ~/react-native
+    docker:
+      - image: circleci/node:8
+    steps:
+      - checkout
+      - run: npm install --no-package-lock
+      - run:
+          name: Building JS bundle for instrumentation tests
+          command: node local-cli/cli.js bundle --max-workers 1 --platform android --dev true --entry-file ReactAndroid/src/androidTest/js/TestBundle.js --bundle-output ReactAndroid/src/androidTest/assets/AndroidTestBundle.js
+      - persist_to_workspace:
+          root: ReactAndroid/src/androidTest/assets/
+          paths:
+            - AndroidTestBundle.js
+
+  test-android:
+    working_directory: ~/react-native
+    docker:
+      - image: circleci/openjdk:8-jdk
+    environment:
+      - TERM: "dumb"
+      - ADB_INSTALL_TIMEOUT: 10
+      - GRADLE_OPTS: '-Dorg.gradle.jvmargs="-Xmx512m -XX:+HeapDumpOnOutOfMemoryError"'
+    steps:
+      - checkout
+      - attach_workspace:
+          at: ReactAndroid/src/androidTest/assets/
+      - run: |
+          if [[ ! -e ReactAndroid/src/androidTest/assets/AndroidTestBundle.js ]]; then
+            echo "JS bundle missing, verify build-js-bundle step"; exit 1
+          else
+            echo "JS bundle found.";
+          fi
+# dependencies
+      - run: echo 'export PATH=~/$CIRCLE_PROJECT_REPONAME/gradle-2.9/bin:/home/circleci/buck/bin:$PATH'
+      - run:
+          name: Setting up Buck and Android dependencies
+          command: |
+            if [[ ! -e /home/circleci/buck ]]; then
+              git clone https://github.com/facebook/buck.git /home/circleci/buck;
+            fi
+            cd /home/circleci/buck && ant
+            buck --version
+            source scripts/circle-ci-android-setup.sh && getAndroidSDK
+            buck fetch ReactAndroid/src/test/java/com/facebook/react/modules
+            buck fetch ReactAndroid/src/main/java/com/facebook/react
+            buck fetch ReactAndroid/src/main/java/com/facebook/react/shell
+            buck fetch ReactAndroid/src/test/...
+            buck fetch ReactAndroid/src/androidTest/...
+            ./gradlew :ReactAndroid:downloadBoost :ReactAndroid:downloadDoubleConversion :ReactAndroid:downloadFolly :ReactAndroid:downloadGlog :ReactAndroid:downloadJSCHeaders
+      - run:
```

Comment:
How about splitting the dependencies step into two separate ones (getting the Android SDK and getting Buck and the RN deps), and then prelaunching the emulator as soon as we've fetched the SDK?

I also suspect that caching some of the Android SDK directories (something to do with ANDROID_HOME or ANDROID_SDK_ROOT, probably need to SSH into the box and poke around to figure out what to cache) would speed up Android builds a good amount without making them flakier, just because developers rarely need to clear out ANDROID_HOME due to bad caching bugs. That said right now I do like this more than creating a Docker image with the Android emulator, really nice not to need to update Docker images <3

Reply:
Good idea on splitting (and actually I did this just now before reading this).

I do want to set up caching, but I'll probably get to it once we have a full successful run. I don't want to pollute the cache or introduce side effects until I'm sure everything is working as expected.

I wish GitHub would not hide review comments automatically with any new commit, as I want to get back to this piece of feedback later on.

Reply:
I made a separate issue to keep track of a couple extra things like Android SDK caching for later: https://github.com/facebook/react-native/issues/15855